### PR TITLE
Expose response body in Python client similar to request body in JS endpoints

### LINKED ISSUE #1584: Expose response body in Python client similar to request body in JS endpoints
Follow-up from https://github.com/microsoft/CCF/pull/1575.

In JS endpoints, the request body is a `Body` object which is similar to https://developer.mozilla.org/en-US/docs/Web/API/Body, except that it has a synchronous API and only offers `.text()`, `.json()`, and `.arrayBuffer()`.

In the Python client, for HTTP, the response content type determines the type that is returned, see `Response.from_requests_response()`, and for WS, given the lack of headers, an attempt is made to parse JSON with fall-back to return `bytes`.

It probably makes sense to align the handling of response bodies in the Python client to how JS endpoints handle request bodies. This gives the developer more control over how they access the body.

### PR DESCRIPTION
**Pull request by letmaik**
Fixes #1584.

Start reviewing at python/ccf/clients.py.

@achamayou @jumaffre Are there other places except tests that use the python client? Docs?

### CODE REVIEW

**Comment by jumaffre** [comment]
LGTM! Thanks for updating this 👍 

It seems that the `/node/tx` endpoint is not too happy:
```
AssertionError: /node/tx should return a JSON object
```

Also, I believe you will also need to update `tutorial.py` in `python/`. To run `tutorial.py` manually (it is only run as part of the daily CI), run `start_test_network.sh ... --network-info-file some_file` and then in a separate terminal `tutorial.py some_file`

**Review by jumaffre — APPROVED** [review]

File: python/ccf/clients.py [hunk 1]
```diff
@@ -81,6 +82,60 @@ def makefile(self, *args, **kwargs):
...
+    def json(self) -> Any:
+        pass
+
+    def __len__(self):
+        return len(self.data())
+
+    def __str__(self):
+        try:
+            return self.text()
+        except UnicodeDecodeError:
```
Comment (achamayou) [inline python/ccf/clients.py:104]: I don't think we should catch that, if a user asks for text, they should get the UnicodeDecodeError and handle it how they want, this is making it very difficult for them to do so.

Comment (letmaik) [reply python/ccf/clients.py:104]: I don't think `str(..)` is ever supposed to throw?

Comment (achamayou) [reply python/ccf/clients.py:104]: No? I can't find a reference either way, this isn't mentioned in https://docs.python.org/3/reference/datamodel.html#object.__str__

At least in 2.x, it definitely used to happen on encoding failures, but that particular behaviour is gone from 3.x.

Could we default to __repr__ on exception perhaps?

**Review by achamayou — APPROVED** [review]

**Comment by achamayou** [comment]
@letmaik this is part of the public interface of the ccf python package, so I think we ought to mention the change in release notes.

**Comment by letmaik** [comment]
I'm running the full test suite locally now.
I noticed that /gov/read returns a list of integers for binary data, instead of the raw binary data. Is that some current limitation?

**Comment by achamayou** [comment]
nlohmann::json serialises std::vector<uint8_t> (our internal representation for most raw buffers) as a list of numbers in json: https://github.com/microsoft/CCF/issues/278

File: python/ccf/clients.py (hunk 1 above)
Comment (letmaik) [reply python/ccf/clients.py:104]: I added the repr fallback.

**Comment by letmaik** [comment]
> 
> 
> nlohmann::json serialises std::vector<uint8_t> (our internal representation for most raw buffers) as a list of numbers in json: #278

Alright, I'll deal with it in the test but I'll open an issue as that's weird.

**Pull request merged**